README.md
  
# Kodi Shield [![License](https://img.shields.io/github/license/robweber/kodi-shield)](https://github.com/robweber/kodi-shield/blob/master/LICENSE) 

This is a PHP application that can be used to generate badges for your Kodi addon using the [Shields.io JSON endpoint](https://shields.io/endpoint) syntax. You can include these badges on your GitHub, or other pages, so people can see information about your addon at glance. The two badges, described below, are Compatible Kodi Version and Total Downloads. 


## Install
You can install this script on any webserver that runs PHP using composer. It does use the Simple XML library for reading in the ```addon.xml``` file of your project so the correct Kodi version can be determined. A sample Apache ```.htaccess``` file using mod_rewrite is given, for other server configurations [PRs are accepted](https://github.com/robweber/kodi-shield/pulls)!

In the ```src/index.php``` file the following variables needed to be modified to fit your server setup: 

```

//MODIFY THESE TO MATCH YOUR SITUATION
$domainPath = 'http://yourdomain.com' //set this to the full path of your server, can be an IP address if no domain name
$basePath = '/kodi-shield'; //set this to the base path on your server

```

Once running you can load the default homepage to see an area where you can preview badges based on a real-time form control. 

## Usage
This script uses the [Shields.io JSON endpoint](https://shields.io/endpoint) to return the JSON needed to generate the badge. Read the information on the Shields.io website for more information on how this works. An example of using the badge within the markdown file of your project could be: 

```
![Kodi Version](https://img.shields.io/endpoint?url=https%3A%2F%2Fdomain.com%2Fkodi-shield%2F/version/%2Frobweber%2Fxbmcbackup)
```

### Compatible Kodi Versions

This badge will show the compatible versions of Kodi for a particular addon. This is calculated by reading the addon.xml file in the addon repository on GitHub.

When queried the application will pull the ```addon.xml``` file of your addon from it's repository and calculate the compatible Kodi versions based on the official [Kodi compatibility matrix](https://kodi.wiki/view/Addon.xml#Dependency_versions). 

__Please note: this script assumes your project is hosted on Github and is the only project in a given repo, with the addon.xml file in the root of the project folder. I've noticed this is the setup for most Kodi addons__ 

The syntax for the endpoint is as follows:

```
/kodi-shield/version/:username/:repo/:branch/:shownames/:currentonly
```

* username __required__ - your Github username
* repo __required__ - the name of the repository for your addon
* branch - the branch name, master is assumed by default
* shownames - a true/false value on if the codenames for each Kodi version should also be shown
* currentonly - another true/false value. By default all compatible Kodi versions are shown, this shows only the most current supported version. 

### Total Downloads

This badge will show the total downloads from the Official Kodi Repo for a particular addon version. This uses stats from Kodi's [mirrorbits]([https://github.com/etix/mirrorbits](https://github.com/etix/mirrorbits)) system. 

The syntax for the endpoint is as follows:

```
/kodi-shield/downloads/:kodi_repo/:addon_id/:addon_version
```

* kodi_repo __required__ - the addon repository that this addon was submitted to. This is _not_ always the same as the Kodi version it is running on. 
* addon_id __required__ - the exact addon ID as specified in your ```addon.xml``` file. 
* addon_version __required__ - the addon version you want stats for. This must match what you see in the Kodi repo to get accurate information. 


## Examples

### Default Compatible Versions Badge 

```
![Kodi Version](https://img.shields.io/endpoint?url=https%3A%2F%2Fdomain.com%2Fkodi-shield%2Fversion%2Frobweber%2Fxbmcbackup)
```

![Kodi Version](https://img.shields.io/endpoint?url=https%3A%2F%2Fweberjr.com%2Fkodi-shield%2Fversion%2Frobweber%2Fxbmcbackup)

###  Compatible Versions Badge with code names:
```
![Kodi Version](https://img.shields.io/endpoint?url=https%3A%2F%2Fdomain.com%2Fkodi-shield%2Fversion%2Frobweber%2Fxbmcbackup%2Fmaster%2Ftrue)
```
![Kodi Version](https://img.shields.io/endpoint?url=https%3A%2F%2Fweberjr.com%2Fkodi-shield%2Fversion%2frobweber%2Fxbmcbackup%2Fmaster%2Ftrue)



### Total Downloads Badge

```
![Kodi Version](https://img.shields.io/endpoint?url=https%3A%2F%2Fdomain.com%2Fkodi-shield%2Fdownloads%2Fmatrix%2Fscript.xbmcbackup%2F1.6.2)
```

![Kodi Version](https://img.shields.io/endpoint?url=https%3A%2F%2Fweberjr.com%2Fkodi-shield%2Fdownloads%2Fmatrix%2Fscript.xbmcbackup%2F1.6.2)

CHANGELOG.md
# Changelog

## [Version 2.5](https://github.com/robweber/kodi-shield/compare/v2.0...v2.5)

### Added
- created an endpoint /downloads that will give total downloads for a specific addon version in a Kodi repo
- added Nexus library comparisons

### Changed
- modifed the url path to generate the badge, it is now ```/version/```
- updated the landing page to include previews for both badge types
- updated placeholder values to show examples

## [Version 2.0](https://github.com/robweber/kodi-shield/compare/v1.0...v2.0)

### Added
- use the [Slim PHP framework](https://www.slimframework.com/) instead of just a basic PHP file. will help expand this to include a few other pages
- added a landing page with preview area
- configurable variables to set domain and path

### Changed
- changed index.php to use Slim Framework, helps with argument parsing

## [Version 1.0](https://github.com/robweber/kodi-shield/compare/v0.1...v1.0)

### Added
- added .htaccess file to redirect all input to index.php
- added use of url path params instead of GET query params
- badge shows red if addon.xml cannot be loaded or parsed, otherwise blue unknown
- when using currentonly param an additional >= modifier is set to show this is the minimum version
- added check if import version actually exists, show unknown if it doesn't
- added imports for other xbmc.* libs

### Changed
- no longer use GET query params (?username=name) and instead use positional params as part of url (/:username/:repo)
- modified README with updated instructions

### Removed
- removed kodi_shield.php for index.php

## [Version 0.1](https://github.com/robweber/kodi-shield/commits/v0.1)

### Added

- added basic README document with some usage examples
- added a changelog document to record changes to this project, format based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
- added first working version. Currently only works for xbmc.python addons, others to be added later


kodi_shield.php
<?php

/**
Kodi Shield Generator

This script will generate the JSON needed to format a badge based on the Shields.io JSON Endpoint specification for Kodi addons. The goal is to show a badge that lists all compatible versions of Kodi for a given addon. 

Query Params: 

* username __required__ - your Github username
* repo __required__ - the name of the repository for your addon
* branch - the branch name, master is assumed by default
* shownames - a true/false value on if the codenames for each Kodi version should also be shown
* currentonly - another true/false value. By default all compatible Kodi versions are shown, this shows only the most current supported version. 

Version: 0.1
Author: Rob Weber
Source: https://github.com/robweber/kodi-shield
License: GPL-3.0 https://github.com/robweber/kodi-shield/blob/master/LICENSE

*/

//valid kodi imports
$kodiImports = array('xbmc.python','xbmc.gui','xbmc.json','xbmc.metadata','xbmc.addon');

//mappings of imports to kodi versions
$kodiNames = array('13.x'=>'Gotham','14.x'=>'Helix','15.x'=>'Isengard','16.x'=>'Jarvis','17.x'=>'Krypton','18.x'=>'Leia');
$kodiMatrix = array('xbmc.python'=>array('2.14.0'=>array('13.x','14.x','15.x','16.x','17.x','18.x'),'2.19.0'=>array('14.x','15.x','16.x','17.x','18.x'),'2.20.0'=>array('15.x','16.x','17.x','18.x'),'2.24.0'=>array('16.x','17.x','18.x'),'2.25.0'=>array('17.x','18.x'),'2.26.0'=>array('18.x')));


$jsonOutput = array('schemaVersion'=>1,'label'=>'kodi version','message'=>'unknown','color'=>'blue');
$validImport = findImport($kodiImports);

if($validImport != null)
{
    $versions = $kodiMatrix[(string)$validImport['addon']][(string)$validImport['version']];

    //check if we should only show the most current
    if(isset($_GET['currentonly']) && $_GET['currentonly'] == 'true')
    {
        $versions = array_slice($versions,0,1);
    }

    //check if we should show versions and names
    if(!isset($_GET['shownames']) || $_GET['shownames'] == 'false')
    {
        $jsonOutput['message'] = implode(', ',$versions);
    }
    else
    {
        $names = array();
        foreach($versions as $aVersion)
        {
            $names[] = sprintf('%s %s',$aVersion,$kodiNames[$aVersion]);
        }

        $jsonOutput['message'] = implode(', ', $names);
    }
}

echo json_encode($jsonOutput);

//finds the first kodi import statement in the addon.xml file
function findImport($kodiImports){
    $validImport = null;

    //we need the user and repo at minimum
    if(isset($_GET['username']) && isset($_GET['repo']))
    {
        //if no branch, assume master
        $branch = 'master';
        if(isset($_GET['branch']))
        {
            $branch = $_GET['branch'];
        }

        //we need the usern, repo, and branch to pull the addon.xml file from
        $repoUrl = sprintf('https://raw.githubusercontent.com/%s/%s/%s/addon.xml',$_GET['username'],$_GET['repo'],$branch);

        $xml = simplexml_load_file($repoUrl);

        foreach($xml->requires->import as $anImport){

            if(in_array((string)$anImport['addon'],$kodiImports))
            {
                $validImport = $anImport;
                break; //we found one
            }
        }
    }

    return $validImport;
}

?>


src/index.php
<?php

/**
Kodi Shield Generator

This script will generate the JSON needed to format a badge based on the Shields.io JSON Endpoint specification for Kodi addons. The goal is to show a badge that lists all compatible versions of Kodi for a given addon.

Url Params:

Example: /kodi-shield/:username/:repo/:branch

Params are positional so using an optional one means you must set the ones before it.

* username (required) - your Github username
* repo (required) - the name of the repository for your addon
* branch - the branch name, master is assumed by default
* shownames - a true/false value on if the codenames for each Kodi version should also be shown
* currentonly - another true/false value. By default all compatible Kodi versions are shown, this shows only the most current supported version.

Version: 2.0
Author: Rob Weber
Source: https://github.com/robweber/kodi-shield
License: GPL-3.0 https://github.com/robweber/kodi-shield/blob/master/LICENSE

*/
use \Psr\Http\Message\ServerRequestInterface as Request;
use \Psr\Http\Message\ResponseInterface as Response;
use \Slim\App as App;

require __DIR__ . '/../vendor/autoload.php';

//valid kodi imports
$kodiImports = array('xbmc.python','xbmc.gui','xbmc.json','xbmc.metadata','xbmc.addon');
//mappings of imports to kodi versions
$kodiNames = array('13.x'=>'Gotham','14.x'=>'Helix','15.x'=>'Isengard','16.x'=>'Jarvis','17.x'=>'Krypton','18.x'=>'Leia','19.x'=>'Matrix','20.x'=>"Nexus","21.x"=>"Omega");
$kodiMatrix = array('xbmc.python'=>array('2.14.0'=>array('13.x','14.x','15.x','16.x','17.x','18.x'),'2.19.0'=>array('14.x','15.x','16.x','17.x','18.x'),'2.20.0'=>array('15.x','16.x','17.x','18.x'),
                                         '2.24.0'=>array('16.x','17.x','18.x'),'2.25.0'=>array('17.x','18.x'),'2.26.0'=>array('18.x'),'3.0.0'=>array('19.x',"20.x"), '3.0.1'=>array('20.x')),
                   'xbmc.gui'=>array('5.0.1'=>array('13.x','14.x','15.x'),'5.3.0'=>array('14.x','15.x'),'5.9.0'=>array('15.x','16.x','17.x','18.x'),'5.10.0'=>array('16.x','17.x','18.x'),
                                     '5.12.0'=>array('17.x','18.x'),'5.14.0'=>array('18.x',"19.x"),'5.15.0'=>array("19.x","20.x"), "5.16.0"=>array("20.x")),
                   'xbmc.json'=>array('6.6.0'=>array('13.x','14.x','15.x','16.x','17.x','18.x',"19.x","20.x"),'6.20.0'=>array('14.x','15.x','16.x','17.x','18.x',"19.x","20.x"),'6.25.1'=>array('15.x','16.x','17.x','18.x','19.x','20.x'),
                                      '6.32.4'=>array('16.x','17.x','18.x','19.x','20.x'),'7.0.0'=>array('17.x','18.x','19.x','20.x'),'9.7.2'=>array('18.x','19.x','20.x'),"11.2.0"=>array("19.x", "20.x"),
                                      "13.0.0"=>array("20.x")),
		                'xbmc.metadata'=>array('2.1.0'=>array('13.x','14.x','15.x','16.x','17.x','18.x','19.x','20.x')),
		                'xbmc.addon'=>array('13.0.0'=>array('13.x','14.x','15.x','16.x','17.x','18.x','19.x','20.x'),'14.0.0'=>array('14.x','15.x','16.x','17.x','18.x','19.x','20.x'),
                                        '15.0.0'=>array('15.x','16.x','17.x','18.x','19.x','20.x'),'16.0.0'=>array('16.x','17.x','18.x','19.x','20.x'),'17.0.0'=>array('17.x','18.x','19.x','20.x'),
                                        '17.9.910'=>array('18.x','19.x','20.x'),'19.1.0'=>array('19.x','20.x'),'20.90.101'=>array('20.x')));

//MODIFY THESE TO MATCH YOUR SITUATION
$domainPath = "https://weberjr.com";
$basePath = '/kodi-shield'; //set this if the basePath changes


$app = new App();

// Get container
$container = $app->getContainer();

// Register component on container
$container['view'] = function ($container) {
    $view = new \Slim\Views\Twig('templates', [
        'cache' => False
    ]);

    // Instantiate and add Slim specific extension
    $router = $container->get('router');
    $uri = \Slim\Http\Uri::createFromEnvironment(new \Slim\Http\Environment($_SERVER));
    $view->addExtension(new \Slim\Views\TwigExtension($router, $uri));

    return $view;
};

//DEFAULT ROUTE
$app->get($basePath . '/', function(Request $request, Response $response, $urlParams) use($basePath, $domainPath) {
    return $this->view->render($response, 'index.html', [
        'basePath' => $basePath,
        'urlRoot' => $domainPath . $basePath
    ]);
});

// ROUTE FOR /version endpoint
$app->get($basePath . '/version/{username}/{repo}[/{branch}[/{shownames}[/{currentonly}]]]', function (Request $request, Response $response, $urlParams) use($kodiImports,$kodiNames,$kodiMatrix) {

    $jsonOutput = array('schemaVersion'=>1,'label'=>'kodi version','message'=>'unknown','color'=>'blue');
    $validImport = findImport($urlParams,$kodiImports);

	if($validImport != null)
	{
		if(array_key_exists((string)$validImport['version'],$kodiMatrix[(string)$validImport['addon']]))
		{
			$versions = $kodiMatrix[(string)$validImport['addon']][(string)$validImport['version']];
			//add version names if wanted
			if(isset($urlParams['shownames']) && $urlParams['shownames'] == 'true')
			{
				$names = array();
				foreach($versions as $aVersion)
				{
					$names[] = sprintf('%s %s',$aVersion,$kodiNames[$aVersion]);
				}
				$versions = $names;
			}
			//create display message
			$message = '';
			//if only the most current get the last in array
			if(isset($urlParams['currentonly']) && $urlParams['currentonly'] == 'true')
			{
				$message = sprintf('%s',$versions[count($versions) -1]);
			}
			else
			{
				$message = implode(', ',$versions);
			}
			$jsonOutput['message'] = $message;
		}
	}
	else
	{
		$jsonOutput['color'] = 'red';
		$jsonOutput['message'] = 'addon.xml error';
	}

		$response->getBody()->write(json_encode($jsonOutput));
		return $response->withHeader('Content-Type','application/json');
	});

//ROUTE FOR /downloads ENDPOINT
$app->get($basePath . '/downloads/{repo}/{addon_name}/{addon_version}', function (Request $request, Response $response, $urlParams) {

	$jsonOutput = array('schemaVersion'=>1,'label'=>sprintf('v%s downloads',$urlParams['addon_version']),'message'=>'unknown','color'=>'green');

	//construct the stats url
	$statUrl = sprintf('http://mirrors.kodi.tv/addons/%s/%s/%s-%s.zip?stats',$urlParams['repo'],$urlParams['addon_name'],$urlParams['addon_name'],$urlParams['addon_version']);

	//if the addon version or name is unknown you just get 0 for everything
	$stats = json_decode(file_get_contents($statUrl));

	$totalDownloads = sprintf('%d', $stats->Total);

	if($stats->Total/1000 > 1)
	{
		$totalDownloads = sprintf('%sK', number_format($stats->Total/1000,1));
	}

	if($stats->Total/1000000 > 1)
	{
		$totalDownloads = sprintf('%sM', number_format($stats->Total/1000000,1));
	}

	$jsonOutput['message'] = $totalDownloads;

	$response->getBody()->write(json_encode($jsonOutput));
	return $response->withHeader('Content-Type','application/json');

	});

$app->run();


//function to find the module imported
function findImport($urlParams,$kodiImports){
    $validImport = null;
    //we need the user and repo at minimum
    if(isset($urlParams['username']) && isset($urlParams['repo']))
    {
        //if no branch, assume master
        $branch = 'master';
        if(isset($urlParams['branch']))
        {
            $branch = $urlParams['branch'];
        }
        //we need the usern, repo, and branch to pull the addon.xml file from
        $repoUrl = sprintf('https://raw.githubusercontent.com/%s/%s/%s/addon.xml',$urlParams['username'],$urlParams['repo'],$branch);
        $xml = simplexml_load_file($repoUrl);
        if($xml !== False)
        {
            foreach($xml->requires->import as $anImport){
                if(in_array((string)$anImport['addon'],$kodiImports))
                {
                    $validImport = $anImport;
                    break; //we found one
                }
            }
        }
    }
    return $validImport;
}

?>
